### PR TITLE
Rtl bfloat

### PR DESCRIPTION
**Pull request by leonardt**
Depends on https://github.com/phanrahan/magma/pull/370 and https://github.com/phanrahan/mantle/pull/143

RTL compilation now works (doesn't crash) for the float ops, however they have not been tested in simulation yet.

### CODE REVIEW

**Comment by leonardt** [comment]
@alexcarsello I had to make some changes to the float op implementations so they matched up with the latest hwtypes/magma. Can you review them to make sure I didn't introduce a bug? The existing tests do pass (they've been unskipped)

**Comment by leonardt** [comment]
My changes were mainly to use the `[width]` syntax, however there were some cases where the widths were not matching up, so I had to change some of the widths (e.g. from 8 to 16). Mainly I'd like a second set of eyes on whether the places where the widths are explicit look okay (i.e. we're not truncating values when we shouldn't be).

**Comment by leonardt** [comment]
Float ops have now been tested on kiwi using a single set of input values!

Depends on https://github.com/leonardt/fault/pull/70 and https://github.com/rdaly525/coreir/pull/713

File: lassen/sim.py [hunk 1]
```diff
@@ -98,80 +99,99 @@ def alu(inst:Inst, a:Data, b:Data, d:Bit) -> (Data, Bit, Bit, Bit, Bit,
...
+        elif alu == ALU.FP_mult:
+            a = BFloat16(a)
+            b = BFloat16(b)
+            res = a * b
+            res_p = Bit(0)
+        elif alu == ALU.FGetMant:
+            res, res_p = (a & 0x7F), Bit(0)
+        elif alu == ALU.FAddIExp:
+            sign = BitVector[16]((a & 0x8000))
+            exp = BitVector[8](((a & 0x7F80)>>7)[0:8])
```
Comment (alexcarsello) [inline lassen/sim.py]: not really on you , but this could be written much more simply as 
`exp = a[7:15]`

File: lassen/sim.py [hunk 2]
```diff
@@ -98,80 +99,99 @@ def alu(inst:Inst, a:Data, b:Data, d:Bit) -> (Data, Bit, Bit, Bit, Bit,
...
+            # Augassign not supported by magma yet
+            # exp += SInt[8](b[0:8])
+            # exp_check += SInt[9](b[0:9])
+            exp_shift = BitVector[16](exp)
+            exp_shift = exp_shift << 7
+            mant = BitVector[16]((a & 0x7F));
+            res, res_p = (sign | exp_shift | mant), (exp_check > 255)
+        elif alu == ALU.FSubExp:
+            signa = BitVector[16]((a & 0x8000))
+            expa = BitVector[8](((a & 0x7F80)>>7)[0:8])
```
Comment (alexcarsello) [inline lassen/sim.py]: same as comment above

**Comment by leonardt** [comment]
Implemented your changes with the simple slicing syntax (had to push a change to magma's lassen-bfloat branch which properly preserved the type in the slice `__getitem__` logic for magma Bits types). Feel free to push any other changes to this branch that you'd like. The RTL tests are working locally for me now (fixed an issue with stubbing out the floating point ops), let's see how they go on travis.

**Comment by alexcarsello** [comment]
there were a few more places where slice should've been used that I didn't mention in my review... just pushed those changes.

**Pull request merged**